presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java

```diff
@@ -450,56 +452,113 @@ public void testReadWriteNDVsComplexTypes()
-        assertEquals(getColumnStatisticsFor(stats, "col0").getDistinctValuesCount(), Estimate.of(2.0));
-        assertEquals(getColumnStatisticsFor(stats, "col1").getDistinctValuesCount(), Estimate.of(2.0));
-        assertEquals(getColumnStatisticsFor(stats, "col2").getDistinctValuesCount(), Estimate.of(2.0));
-        assertEquals(getColumnStatisticsFor(stats, "col3").getDistinctValuesCount(), Estimate.unknown());
+        assertEquals(columnStatsFor(stats, "col0").getDistinctValuesCount(), Estimate.of(2.0));
+        assertEquals(columnStatsFor(stats, "col1").getDistinctValuesCount(), Estimate.of(2.0));
+        assertEquals(columnStatsFor(stats, "col2").getDistinctValuesCount(), Estimate.of(2.0));
+        assertEquals(columnStatsFor(stats, "col3").getDistinctValuesCount(), Estimate.unknown());
 
         assertUpdate("DROP TABLE test_stat_ndv_complex");
     }
 
+    @Test
+    public void testNDVsAtSnapshot()
+    {
+        assertUpdate("CREATE TABLE test_stat_snap (col0 int, col1 varchar)");
+        assertTrue(getQueryRunner().tableExists(getSession(), "test_stat_snap"));
+        assertTableColumnNames("test_stat_snap", "col0", "col1");
+        assertEquals(getTableSnapshots("test_stat_snap").size(), 0);
+
+        assertQuerySucceeds("ANALYZE test_stat_snap");
+        assertUpdate("INSERT INTO test_stat_snap VALUES (0, '0')", 1);
+        assertQuerySucceeds("ANALYZE test_stat_snap");
+        assertUpdate("INSERT INTO test_stat_snap VALUES (1, '1')", 1);
+        assertQuerySucceeds("ANALYZE test_stat_snap");
+        assertUpdate("INSERT INTO test_stat_snap VALUES (2, '2')", 1);
+        assertQuerySucceeds("ANALYZE test_stat_snap");
+        assertUpdate("INSERT INTO test_stat_snap VALUES (3, '3')", 1);
+        assertQuerySucceeds("ANALYZE test_stat_snap");
+        assertEquals(getTableSnapshots("test_stat_snap").size(), 4);
+
+        List<Long> snaps = getTableSnapshots("test_stat_snap");
+        for (int i = 0; i < snaps.size(); i++) {
+            TableStatistics statistics = getTableStats("test_stat_snap", Optional.of(snaps.get(i)));
+            // assert either case as we don't have good control over the timing of when statistics files are written
+            ColumnStatistics col0Stats = columnStatsFor(statistics, "col0");
+            ColumnStatistics col1Stats = columnStatsFor(statistics, "col1");
+            System.out.printf("distinct @ %s count col0: %s%n", snaps.get(i), col0Stats.getDistinctValuesCount());
+            final int idx = i;
+            assertEither(
+                    () -> assertEquals(col0Stats.getDistinctValuesCount(), Estimate.of(idx)),
+                    () -> assertEquals(col0Stats.getDistinctValuesCount(), Estimate.of(idx + 1)));
+            assertEither(
+                    () -> assertEquals(col1Stats.getDistinctValuesCount(), Estimate.of(idx)),
+                    () -> assertEquals(col1Stats.getDistinctValuesCount(), Estimate.of(idx + 1)));
```

Comment:
I think in this test scenario, the statistics analyzed exactly on target snapshot would always be chosen. As I understand, only when there do not exist statistics on target snapshot need we  to choose a closer statistics.

Reply:
Another test added in 4345aa7a296374b5105a287c3b06d9b53bd156f7 which makes use of the distance calculation and weight parameter